source/common/upstream/cluster_manager_impl.cc

```diff
@@ -336,28 +338,72 @@ void ClusterManagerImpl::onClusterInit(Cluster& cluster) {
   }
 }
 
-bool ClusterManagerImpl::addOrUpdatePrimaryCluster(const envoy::api::v2::Cluster& cluster) {
+bool ClusterManagerImpl::addOrUpdateCluster(const envoy::api::v2::Cluster& cluster) {
   // First we need to see if this new config is new or an update to an existing dynamic cluster.
   // We don't allow updates to statically configured clusters in the main configuration.
   const std::string cluster_name = cluster.name();
-  auto existing_cluster = primary_clusters_.find(cluster_name);
-  if (existing_cluster != primary_clusters_.end() &&
-      (!existing_cluster->second.added_via_api_ ||
-       existing_cluster->second.config_hash_ == MessageUtil::hash(cluster))) {
+  const auto existing_active_cluster = active_clusters_.find(cluster_name);
+  const auto existing_warming_cluster = warming_clusters_.find(cluster_name);
+  const uint64_t new_hash = MessageUtil::hash(cluster);
+  if ((existing_active_cluster != active_clusters_.end() &&
+       existing_active_cluster->second->blockUpdate(new_hash)) ||
+      (existing_warming_cluster != warming_clusters_.end() &&
+       existing_warming_cluster->second->blockUpdate(new_hash))) {
     return false;
   }
 
-  if (existing_cluster != primary_clusters_.end()) {
-    init_helper_.removeCluster(*existing_cluster->second.cluster_);
+  if (existing_active_cluster != active_clusters_.end() ||
+      existing_warming_cluster != warming_clusters_.end()) {
+    init_helper_.removeCluster(*existing_active_cluster->second->cluster_);
```

Comment:
How is this valid if we get here due to the warming_cluster_ clause in the conditional?

Reply:
There is a guard that protects it. I will make it more clear.